servicetalk-transport-api/src/main/java/io/servicetalk/transport/api/DefaultHostAndPort.java

```diff
@@ -62,4 +135,126 @@ public boolean equals(Object o) {
+                    lastColon = i;
+                    break;
+                default:
+                    if (!isValidateIPv6Digit(c)) {
+                        boolean shouldInsertRemainder = false;
+                        if (c == '.') { // check for IPv4 mapped
+                            final int zoneIdIndex = rawIp.lastIndexOf('%');
+                            validateIPv4(rawIp, lastColon + 1, zoneIdIndex < 0 ? rawIp.length() : zoneIdIndex);
+                            shouldInsertRemainder = true;
+                        } else if (c == '%') { // check for IPv6 zone id
+                            // no constraints enforced on zone id
+                            shouldInsertRemainder = true;
+                        }
+                        if (shouldInsertRemainder) {
+                            compressedIPv6Builder.append(c);
+                            ++i;
+                            for (; i < rawIp.length(); ++i) {
+                                compressedIPv6Builder.append(rawIp.charAt(i));
+                            }
+                            break;
+                        } else {
+                            throw new IllegalArgumentException("Invalid IPv6 address[" + i + "]=" + c);
+                        }
+                    }
+                    // https://datatracker.ietf.org/doc/html/rfc5952#section-4.2.3
+                    // if there is a tie in the longest length, we must choose the first to compress.
+                    if (zerosEnd > 0 && zerosCount > longestZerosCount) {
+                        longestZerosCount = zerosCount;
+                        longestZerosBegin = zerosBegin;
+                        longestZerosEnd = zerosEnd;
+                    }
+                    zerosBegin = zerosEnd = -1;
+                    zerosCount = 0;
+                    compressedIPv6Builder.append(c);
+                    break;
+            }
+            prevChar = c;
+        }
+        // https://datatracker.ietf.org/doc/html/rfc5952#section-4.2.2
+        // The symbol "::" MUST NOT be used to shorten just one 16-bit 0 field.
+        if (!isCompressed && longestZerosBegin >= 0 && longestZerosCount > 1) {
+            compressedIPv6Builder.replace(longestZerosBegin, longestZerosEnd, longestZerosBegin == 0 ? "::" : ":");
+        }
+        return compressedIPv6Builder.toString();
+    }
+
+    private static boolean isValidateIPv6Digit(char c) {
+        return (c >= '0' && c <= '9') || (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F');
+    }
+
+    private static void validateIPv4(String inetAddress, int startIndex, int endIndex) {
+        int nibbles = 0;
+        int digits = 0;
+        for (int i = startIndex; i < endIndex; ++i) {
+            final char currChar = inetAddress.charAt(i);
+            if (currChar >= '0' && currChar <= '9') {
+                if (++digits > 3) {
+                    throw new IllegalArgumentException("No more than 3 digits per section expected");
```

Comment:
Consider including a substring of the validating area to let users see what they gave as a parameter

Reply:
I used NetworkUtils for this now and print substring as suggested